File: examples/langchain/langchain_tool_arcade_auth.py
import os

from arcadepy import Arcade
from google.oauth2.credentials import Credentials
from langchain_google_community import GmailToolkit
from langchain_google_community.gmail.utils import (
    build_resource_service,
)
from langchain_openai import ChatOpenAI
from langgraph.prebuilt import create_react_agent

# Get the API key from the environment variable
api_key = os.getenv("ARCADE_API_KEY")

# Initialize the Arcade client
client = Arcade(api_key=api_key)

# Start the authorization process for Gmail
# see all possible gmail scopes here:
# https://developers.google.com/gmail/api/auth/scopes
user_id = "user@example.com"
auth_response = client.auth.start(
    user_id=user_id, provider="google", scopes=["https://www.googleapis.com/auth/gmail.readonly"]
)

# Prompt the user to authorize if not already completed
if auth_response.status != "completed":
    print("Please authorize the application in your browser:")
    print(auth_response.url)

# Wait for the user to complete the authorization process, if necessary...
auth_response = client.auth.wait_for_completion(auth_response)

# Obtain credentials using the authorization context
creds = Credentials(auth_response.context.token)
api_resource = build_resource_service(credentials=creds)

# Initialize the Gmail toolkit with the authorized API resource
toolkit = GmailToolkit(api_resource=api_resource)

# Retrieve the tools from the langchain gmail toolkit
tools = toolkit.get_tools()

# Initialize the language model and create an agent
llm = ChatOpenAI(model="gpt-4o")
agent_executor = create_react_agent(llm, tools)

# Define the user query
example_query = "Read my latest emails and summarize them."

# Execute the agent with the user query
events = agent_executor.stream(
    {"messages": [("user", example_query)]},
    stream_mode="values",
)

# Display the agent's response
for event in events:
    event["messages"][-1].pretty_print()

File: examples/langchain-ts/langchain-tool-arcade-auth.ts
import { Arcade } from "@arcadeai/arcadejs";
import {
	GmailCreateDraft,
	GmailGetMessage,
	GmailGetThread,
	GmailSearch,
	GmailSendMessage,
} from "@langchain/community/tools/gmail";
import type { StructuredTool } from "@langchain/core/tools";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { ChatOpenAI } from "@langchain/openai";

// Initialize Arcade with API key from environment
const arcade = new Arcade();
// Replace this with your application's user ID (e.g. email address, UUID, etc.)
const USER_ID = "user@example.com";

// Start the authorization process for Gmail
// see all possible gmail scopes here:
// https://developers.google.com/gmail/api/auth/scopes
const authResponse = await arcade.auth.start(USER_ID, "google", {
	scopes: ["https://www.googleapis.com/auth/gmail.readonly"],
});

// Prompt the user to authorize if not already completed
if (authResponse.status !== "completed") {
	console.log("Please authorize the application in your browser:");
	console.log(authResponse.url);
}

// Wait for the user to complete the authorization process, if necessary...
const completedAuth = await arcade.auth.waitForCompletion(authResponse);

if (!completedAuth.context?.token) {
	throw new Error("Failed to get authorization token");
}

// Get Gmail tools with credentials
const gmailParam = {
	credentials: {
		accessToken: completedAuth.context.token,
	},
};
const tools: StructuredTool[] = [
	new GmailCreateDraft(gmailParam),
	new GmailGetMessage(gmailParam),
	new GmailGetThread(gmailParam),
	new GmailSearch(gmailParam),
	new GmailSendMessage(gmailParam),
];

// Initialize the language model and create an agent
const llm = new ChatOpenAI({ model: "gpt-4o" });
const agent_executor = createReactAgent({
	llm,
	tools,
});

// Define the user query
const exampleQuery = "Read my latest emails and summarize them.";

// Execute the agent with the user query
const events = await agent_executor.stream(
	{
		messages: [
			{
				role: "user",
				content: exampleQuery,
			},
		],
	},
	{
		streamMode: "values",
	},
);

for await (const event of events) {
	console.log(event.messages[event.messages.length - 1]);
}
